sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java

```diff
@@ -253,7 +255,7 @@ protected AnalyzedRelation visitJoin(Join node, StatementAnalysisContext stateme
         Symbol joinCondition = null;
         if (optCriteria.isPresent()) {
             JoinCriteria joinCriteria = optCriteria.get();
-            if (joinCriteria instanceof JoinOn) {
+            if (joinCriteria instanceof JoinOn || joinCriteria instanceof JoinUsing) {
```

Comment:
How about something like the following to avoid the duplicate instanceof:

```
diff --git a/sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java b/sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java
index 80f718714f..8e3c5da7cf 100644
--- a/sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java
+++ b/sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java
@@ -107,7 +107,6 @@ import javax.annotation.Nullable;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
-import java.util.Iterator;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
@@ -247,48 +246,45 @@ public class RelationAnalyzer extends DefaultTraversalVisitor<AnalyzedRelation,
 
     @Override
     protected AnalyzedRelation visitJoin(Join node, StatementAnalysisContext statementContext) {
-        node.getLeft().accept(this, statementContext);
-        node.getRight().accept(this, statementContext);
+        var leftRel = node.getLeft().accept(this, statementContext);
+        var rightRel = node.getRight().accept(this, statementContext);
 
         RelationAnalysisContext relationContext = statementContext.currentRelationContext();
         Optional<JoinCriteria> optCriteria = node.getCriteria();
         Symbol joinCondition = null;
         if (optCriteria.isPresent()) {
             JoinCriteria joinCriteria = optCriteria.get();
-            if (joinCriteria instanceof JoinOn || joinCriteria instanceof JoinUsing) {
-                final CoordinatorTxnCtx coordinatorTxnCtx = statementContext.transactionContext();
-                ExpressionAnalyzer expressionAnalyzer = new ExpressionAnalyzer(
-                    functions,
-                    coordinatorTxnCtx,
-                    statementContext.convertParamFunction(),
-                    new FullQualifiedNameFieldProvider(
-                        relationContext.sources(),
-                        relationContext.parentSources(),
-                        coordinatorTxnCtx.sessionContext().searchPath().currentSchema()),
-                    new SubqueryAnalyzer(this, statementContext));
-                Expression expr;
-                if (joinCriteria instanceof JoinOn) {
-                    expr = ((JoinOn) joinCriteria).getExpression();
-                } else {
-                    Iterator<QualifiedName> rels = relationContext.sources().keySet().iterator();
-                    QualifiedName left = rels.hasNext() ? rels.next() : null;
-                    QualifiedName right = rels.hasNext() ? rels.next() : null;
-                    if (null == left || null == right) {
-                        throw new IllegalStateException("joining tables are undefined");
-                    }
-                    expr = JoinUsing.toExpression(left, right, ((JoinUsing) joinCriteria).getColumns());
-                }
-                try {
-                    joinCondition = expressionAnalyzer.convert(expr, relationContext.expressionAnalysisContext());
-                } catch (RelationUnknown e) {
-                    throw new RelationValidationException(e.getTableIdents(),
-                        String.format(Locale.ENGLISH,
-                        "missing FROM-clause entry for relation '%s'", e.getTableIdents()));
-                }
+            final CoordinatorTxnCtx coordinatorTxnCtx = statementContext.transactionContext();
+            ExpressionAnalyzer expressionAnalyzer = new ExpressionAnalyzer(
+                functions,
+                coordinatorTxnCtx,
+                statementContext.convertParamFunction(),
+                new FullQualifiedNameFieldProvider(
+                    relationContext.sources(),
+                    relationContext.parentSources(),
+                    coordinatorTxnCtx.sessionContext().searchPath().currentSchema()),
+                new SubqueryAnalyzer(this, statementContext)
+            );
+            final Expression expr;
+            if (joinCriteria instanceof JoinOn) {
+                expr = ((JoinOn) joinCriteria).getExpression();
+            } else if (joinCriteria instanceof JoinUsing) {
+                expr = JoinUsing.toExpression(
+                    leftRel.getQualifiedName(),
+                    rightRel.getQualifiedName(),
+                    ((JoinUsing) joinCriteria).getColumns()
+                );
             } else {
                 throw new UnsupportedOperationException(String.format(Locale.ENGLISH, "join criteria %s not supported",
                     joinCriteria.getClass().getSimpleName()));
             }
+            try {
+                joinCondition = expressionAnalyzer.convert(expr, relationContext.expressionAnalysisContext());
+            } catch (RelationUnknown e) {
+                throw new RelationValidationException(e.getTableIdents(),
+                    String.format(Locale.ENGLISH,
+                    "missing FROM-clause entry for relation '%s'", e.getTableIdents()));
+            }
         }
 
         relationContext.addJoinType(JoinType.values()[node.getType().ordinal()], joinCondition);

```

and to avoid the left / right null cases.